Add get_torrents service to qBittorrent integration

finder39 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a `get_torrent()` service to the qBittorrent inragration

![image](https://github.com/home-assistant/core/assets/518673/3e9990c7-c124-4937-bfcd-aedf689629b7)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/32384

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

finder39 committed:
Upgrade QBittorrent integration to show torrents

This brings the QBittorrent integration to be more in line with the Transmission integration. It updates how the integration is written, along with adding sensors for Active Torrents, Inactive Torrents, Paused Torrents, Total Torrents, Seeding Torrents, Started Torrents.

finder39 committed:
Remove unused stuff

finder39 committed:
Correct name in comments

finder39 committed:
Make get torrents a service with a response

finder39 committed:
Add new sensors

finder39 committed:
remove service

finder39 committed:
Add service with response to get torrents list

This adds a service with a response to be able to get the list of torrents within qBittorrent

finder39 committed:
update

finder39 committed:
update from rebase

finder39 committed:
Update strings.json

finder39 committed:
Update helpers.py

finder39 committed:
Update to satisfy lint

joostlek commented:
@engrbm87 can you also take a look at this? Since I also promoted the use of a service call for transmission, maybe it's an idea to use a similar return schema so it's easier for custom card creators to create a card that works for both integrations.

engrbm87 commented:
> @engrbm87 can you also take a look at this? Since I also promoted the use of a service call for transmission, maybe it's an idea to use a similar return schema so it's easier for custom card creators to create a card that works for both integrations. 

Sure, I will work on something similar for transmission. But I would like to know how the response can be displayed on the frontend (like a markdown card).

finder39 commented:
> > @engrbm87 can you also take a look at this? Since I also promoted the use of a service call for transmission, maybe it's an idea to use a similar return schema so it's easier for custom card creators to create a card that works for both integrations.
> 
> Sure, I will work on something similar for transmission. But I would like to know how the response can be displayed on the frontend (like a markdown card).

I built a custom card that does that for qBittorrent with these changes. I can upload it to github later today

engrbm87 commented:
> > > @engrbm87 can you also take a look at this? Since I also promoted the use of a service call for transmission, maybe it's an idea to use a similar return schema so it's easier for custom card creators to create a card that works for both integrations.
> > 
> > Sure, I will work on something similar for transmission. But I would like to know how the response can be displayed on the frontend (like a markdown card).
> 
> I built a custom card that does that for qBittorrent with this new service. I can upload it to github later today

Will this card work for any service response or is it specific for qBittorent. It would be hard to build a separate card for each integration.

finder39 commented:
> > > > @engrbm87 can you also take a look at this? Since I also promoted the use of a service call for transmission, maybe it's an idea to use a similar return schema so it's easier for custom card creators to create a card that works for both integrations.
> > > 
> > > 
> > > Sure, I will work on something similar for transmission. But I would like to know how the response can be displayed on the frontend (like a markdown card).
> > 
> > 
> > I built a custom card that does that for qBittorrent with this new service. I can upload it to github later today
> 
> Will this card work for any service response or is it specific for qBittorent. It would be hard to build a separate card for each integration.

I mean the card is specifically built for qBittorrent in this case, but its using the standard HA way of calling a service with a response. I built it off of https://github.com/amaximus/transmission-card, but customized it for qbittorrent and its new service call

joostlek commented:
I'm slowly playing with the idea of "standardised service calls". We of course have them for our entity platforms like "climate.set_hvac" or "calendar.get_events". But I think it would benefit some types of integrations as well to have a standard service call. For example public transport integrations or download integrations.

And currently there are no plans afaik to support something like that in core itself, but I'll fish around checking what others think of it and maybe we can implement it.

This doesn't have to hold us back into making 2 "custom" services return in the same data structure

joostlek commented:
And the idea behind it is that we don't create a complete new entity platform for public transport or download integrations, but just keep using these mostly sensor based integrations

finder39 committed:
add func comment

finder39 committed:
fix lint issues

finder39 committed:
another update attempt

finder39 committed:
Fix helpers

joostlek commented:
@tkdrob can you also take a look for Deluge?

finder39 committed:
Remove unneccesary part in services.yaml and add translations

finder39 committed:
Merge branch 'dev' into qbittorrent-add-get-torrents-service

finder39 committed:
Fix return

finder39 committed:
Add tests

finder39 committed:
Fix test

finder39 committed:
Improve tests

finder39 committed:
Merge branch 'dev' into qbittorrent-add-get-torrents-service

finder39 commented:
@engrbm87 Take at look at the [qBittorrent card here](https://github.com/finder39/qbittorrent-card/blob/dev/qbittorrent-card.js). I'm in the process of cleaning it up (and want to rebase off the changes that azamimus made, but it gives a good idea of how to handle this.

The main functions to look at are:
- `_getTorrents`: handles the get_torrents service call from HA
- `_setupGetTorrentsInterval`: handles setting up the interval at which the service call is made
- `updated`: enables initial setup of service call interval and updates it when needed
- `disconnectedCallback`: clears out the interval call when required

engrbm87 commented:
> @engrbm87 Take at look at the [qBittorrent card here](https://github.com/finder39/qbittorrent-card/blob/dev/qbittorrent-card.js). I'm in the process of cleaning it up (and want to rebase off the changes that azamimus made, but it gives a good idea of how to handle this.
> 
> The main functions to look at are:
> 
> * `_getTorrents`: handles the get_torrents service call from HA
> * `_setupGetTorrentsInterval`: handles setting up the interval at which the service call is made
> * `updated`: enables initial setup of service call interval and updates it when needed
> * `disconnectedCallback`: clears out the interval call when required

@finder39 Thanks for sharing the progress of the custom card. I have a couple of ideas for this card. Perhaps you can enable the discussion or issues tab on your repository and we can discuss it there.

finder39 commented:
> @finder39 Thanks for sharing the progress of the custom card. I have a couple of ideas for this card. Perhaps you can enable the discussion or issues tab on your repository and we can discuss it there.

done!

finder39 commented:
> @finder39 Thanks for sharing the progress of the custom card. I have a couple of ideas for this card. Perhaps you can enable the discussion or issues tab on your repository and we can discuss it there.

@engrbm87  Done!

finder39 committed:
Merge branch 'dev' into qbittorrent-add-get-torrents-service

joostlek commented:
Whoops, I should've looked at this

joostlek commented on homeassistant/components/qbittorrent/__init__.py:
```diff
@@ -46,6 +46,23 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
 
     await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
 
+    async def handle_get_torrents(service_call: ServiceCall):
+        coordinator: QBittorrentDataCoordinator = hass.data[DOMAIN][
+            config_entry.entry_id
+        ]
+        items = await coordinator.get_torrents(service_call.data[TORRENT_FILTER])
+        info = format_torrents(items)
+        return {
+            STATE_ATTR_TORRENT_INFO: info,
+        }
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_GET_TORRENTS,
+        handle_get_torrents,
+        supports_response=SupportsResponse.ONLY,
+    )
```
I think our general consensus is now that we want this to be in `async_setup`, otherwise the service is dependent on the config entry.

I am also not sure how this exactly works when you have 2 config entries for qbittorrent

finder39 replied:
Lemme take a look at this and check

joostlek commented on homeassistant/components/qbittorrent/helpers.py:
```diff
@@ -9,3 +12,51 @@ def setup_client(url: str, username: str, password: str, verify_ssl: bool) -> Cl
     return client
+
+
+def seconds_to_hhmmss(seconds):
+    """Convert seconds to HH:MM:SS format."""
+    if seconds == 8640000:
+        return "None"
+
+    minutes, seconds = divmod(seconds, 60)
+    hours, minutes = divmod(minutes, 60)
+    return f"{int(hours):02}:{int(minutes):02}:{int(seconds):02}"
+
+
+def format_unix_timestamp(timestamp):
+    """Format a UNIX timestamp to a human-readable date."""
+    dt_object = datetime.fromtimestamp(timestamp, tz=UTC)
+    formatted_date = dt_object.strftime("%Y-%m-%dT%H:%M:%S%z")
+    return formatted_date
+
+
+def format_progress(torrent):
+    """Format the progress of a torrent."""
+    progress = torrent["progress"]
+    progress = float(progress) * 100
+    progress = f"{progress:.2f}"
+
+    return progress
+
+
+def format_torrents(torrents: list[dict[str, Any]]):
+    """Format a list of torrents."""
+    value = {}
+    for torrent in torrents:
+        value[torrent["name"]] = format_torrent(torrent)
+
+    return value
+
+
+def format_torrent(torrent):
```
Can we completely type this?

joostlek reviewed (dismissed):
I don't know what the test state of QBittorrent is, but would it be possible to add tests for this? We can use snapshots to snapshot the service response, this way the response is visible in the code.

finder39 commented:
> I don't know what the test state of QBittorrent is, but would it be possible to add tests for this? We can use snapshots to snapshot the service response, this way the response is visible in the code.

I don't believe there are many tests. Let me see what it would entail to add some for this

finder39 committed:
Add service with response to get torrents list

This adds a service with a response to be able to get the list of torrents within qBittorrent

finder39 committed:
Make get torrents a service with a response

finder39 committed:
remove service

finder39 committed:
Update to satisfy lint

finder39 committed:
Merge remote-tracking branch 'upstream/dev' into qbittorrent-add-get-torrents-service

finder39 committed:
Fix issue from rebase

finder39 committed:
Add icon for get_torrents service

finder39 committed:
Handle multiple installed integrations

finder39 committed:
fix lint issue

finder39 committed:
Set return types for helper methods

finder39 committed:
Create the service method in async_setup

finder39 committed:
Add CONFIG_SCHEMA

finder39 committed:
Add get_all_torrents service

finder39 committed:
Merge branch 'home-assistant:dev' into qbittorrent-add-get-torrents-service

finder39 committed:
fix lint issues

emontnemery commented on homeassistant/components/qbittorrent/__init__.py:
```diff
@@ -7,23 +7,92 @@
+        device_entry = device_registry.async_get(service_call.data[ATTR_DEVICE_ID])
+
+        if device_entry is None:
+            return
```
This should raise a `ServiceValidationError`

emontnemery commented on homeassistant/components/qbittorrent/__init__.py:
```diff
@@ -7,23 +7,92 @@
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up qBittorrent services."""
+
+    async def handle_get_torrents(service_call: ServiceCall):
```
Please annotate the return type

emontnemery commented on homeassistant/components/qbittorrent/__init__.py:
```diff
@@ -7,23 +7,92 @@
+                entry_id = value
+                break
+        else:
+            return
```
This too should raise a `ServiceValidationError`

emontnemery commented on homeassistant/components/qbittorrent/coordinator.py:
```diff
@@ -33,7 +40,19 @@ def __init__(self, hass: HomeAssistant, client: Client) -> None:
+    async def get_torrents(self, torrent_filter: str) -> list[dict[str, Any]]:
+        """Async method to get QBittorrent torrents."""
+        try:
+            total_torrents = await self.hass.async_add_executor_job(
```
`total_torrents` is a bit weird for something which is a list, `torrents` would be better.

emontnemery commented on homeassistant/components/qbittorrent/coordinator.py:
```diff
@@ -33,7 +40,19 @@ def __init__(self, hass: HomeAssistant, client: Client) -> None:
+                lambda: self.client.torrents(filter=torrent_filter)
+            )
+        except LoginRequired as exc:
+            raise ConfigEntryError("Invalid authentication") from exc
```
I think the service handlers need to catch this and reraise a translated `HomeAssistantError`

finder39 committed:
Add return types and ServiceValidationError(s)

finder39 commented:
The requested changes have been made

finder39 committed:
Fix naming

emontnemery commented on homeassistant/components/qbittorrent/strings.json:
```diff
@@ -48,5 +48,39 @@
+  },
+  "exceptions": {
+    "invalid_device": {
+      "message": "No device with that id was found"
```
Pass in the device_id instead, for a more readable error message:
```suggestion
      "message": "No device with id {device_id} was found"
```

emontnemery commented on homeassistant/components/qbittorrent/coordinator.py:
```diff
@@ -33,7 +40,19 @@ def __init__(self, hass: HomeAssistant, client: Client) -> None:
+                lambda: self.client.torrents(filter=torrent_filter)
+            )
+        except LoginRequired as exc:
+            raise HomeAssistantError(str(exc)) from exc
```
Shouldn't this be translated, or can `LoginRequired` happen in many cases?

finder39 replied:
I handled this how I found other integrations were but i can change this to be a translated string.

I believe this is a very rare occurance as it should only happen if a user changes the username/password of their qBittorrent software after adding to HA. As this is a local program running that would be very rare and there isn't a need to change that for security.

emontnemery commented on homeassistant/components/qbittorrent/helpers.py:
```diff
@@ -10,3 +13,48 @@ def setup_client(url: str, username: str, password: str, verify_ssl: bool) -> Cl
+def format_unix_timestamp(timestamp) -> str:
+    """Format a UNIX timestamp to a human-readable date."""
+    dt_object = datetime.fromtimestamp(timestamp, tz=UTC)
+    return dt_object.strftime("%Y-%m-%dT%H:%M:%S%z")
```
What's the difference between the manual formatting and the output of the `isoformat` method? Why do we format it at all?

finder39 committed:
Update translations

finder39 committed:
Fix tests

emontnemery reviewed (approved):
Thanks, @finder39 👍

MartinHjelmare commented on homeassistant/components/qbittorrent/__init__.py:
```diff
@@ -1,29 +1,111 @@
+        DOMAIN,
+        SERVICE_GET_TORRENTS,
+        handle_get_torrents,
+        supports_response=SupportsResponse.ONLY,
```
I'm missing a service validation schema.

MartinHjelmare commented on homeassistant/components/qbittorrent/const.py:
```diff
@@ -7,6 +7,13 @@
 DEFAULT_URL = "http://127.0.0.1:8080"
 
+STATE_ATTR_TORRENTS = "torrents"
+STATE_ATTR_ALL_TORRENTS = "all_torrents"
```
These are not state attribute. Please rename the constants.

MartinHjelmare commented on homeassistant/components/qbittorrent/const.py:
```diff
@@ -7,6 +7,13 @@
 STATE_SEEDING = "seeding"
 STATE_DOWNLOADING = "downloading"
+
+SERVICE_GET_TORRENTS = "get_torrents"
```
We don't need constants for strings that are only used once.

MartinHjelmare commented on homeassistant/components/qbittorrent/coordinator.py:
```diff
@@ -19,11 +19,18 @@
     def __init__(self, hass: HomeAssistant, client: Client) -> None:
         """Initialize coordinator."""
         self.client = client
+        # self.main_data: dict[str, int] = {}
```
Commented code. Please remove it.

MartinHjelmare commented on homeassistant/components/qbittorrent/coordinator.py:
```diff
@@ -19,11 +19,18 @@
         """Initialize coordinator."""
         self.client = client
+        # self.main_data: dict[str, int] = {}
+        self.total_torrents: dict[str, int] = {}
+        self.active_torrents: dict[str, int] = {}
+        self.inactive_torrents: dict[str, int] = {}
+        self.paused_torrents: dict[str, int] = {}
+        self.seeding_torrents: dict[str, int] = {}
+        self.started_torrents: dict[str, int] = {}
```
These instance attributes are never used. Please remove them.

MartinHjelmare commented on homeassistant/components/qbittorrent/coordinator.py:
```diff
@@ -33,7 +40,21 @@ def __init__(self, hass: HomeAssistant, client: Client) -> None:
             return await self.hass.async_add_executor_job(self.client.sync_main_data)
         except LoginRequired as exc:
-            raise ConfigEntryError("Invalid authentication") from exc
+            raise HomeAssistantError(str(exc)) from exc
```
Why did we change the exception in this PR? That looks unrelated to the new services.

MartinHjelmare commented on homeassistant/components/qbittorrent/helpers.py:
```diff
@@ -10,3 +13,48 @@ def setup_client(url: str, username: str, password: str, verify_ssl: bool) -> Cl
     return client
+
+
+def seconds_to_hhmmss(seconds) -> str:
```
The parameter is missing a type annotation. Please type the whole signature when adding type annotations.

MartinHjelmare commented on homeassistant/components/qbittorrent/strings.json:
```diff
@@ -48,5 +48,42 @@
+      "fields": {
+        "device_id": {
+          "name": "[%key:common::config_flow::data::device%]",
+          "description": "Which service to grab the list from"
```
"grab the list" is too colloquial English and not correctly explained what it's referring to.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
